stock_quant_constraint/models/stock_quant.py

```diff
@@ -0,0 +1,206 @@
+    SELECT
+        sq.product_id, sq.location_id,
+            sq.lot_id, sq.package_id, sq.owner_id,
+            sq.quantity
+    FROM stock_quant sq
+),
+quant_summary AS (
+    SELECT
+        COALESCE(q.product_id, im.product_id, om.product_id) AS product_id,
+            COALESCE(q.location_id, im.location_dest_id, om.location_id) AS location_id,
+            COALESCE(q.lot_id, im.lot_id, om.lot_id) AS lot_id,
+            COALESCE(q.package_id, im.package_id, om.package_id) AS package_id,
+            q.quantity,
+            COALESCE(q.owner_id, im.owner_id, om.owner_id) AS owner_id,
+            im.qty_done AS quantity_in, om.qty_done AS quantity_out,
+            COALESCE(q.quantity, 0) - (COALESCE(im.qty_done, 0)
+                - COALESCE(om.qty_done, 0)) AS quantity_error
+    FROM quants q
+    FULL OUTER JOIN incoming_moves im ON
+    q.product_id = im.product_id AND q.location_id = im.location_dest_id
+    AND (q.lot_id = im.lot_id OR (q.lot_id IS NULL AND im.lot_id IS NULL))
+    AND (q.package_id = im.package_id
+        OR (q.package_id IS NULL AND im.package_id IS NULL))
+    AND (q.owner_id = im.owner_id OR (q.owner_id IS NULL AND im.owner_id IS NULL))
+    FULL OUTER JOIN outgoing_moves om ON
+    q.product_id = om.product_id AND q.location_id = om.location_id
+    AND (q.lot_id = om.lot_id OR (q.lot_id IS NULL AND om.lot_id IS NULL))
+    AND (q.package_id = om.package_id
+        OR (q.package_id IS NULL AND om.package_id IS NULL))
+    AND (q.owner_id = om.owner_id OR (q.owner_id IS NULL AND om.owner_id IS NULL))
+)
+SELECT
+    qs.product_id,
+    qs.location_id,
+    qs.package_id,
+    qs.owner_id,
+    qs.lot_id,
+    COUNT(*) as kount,
+    SUM(quantity) AS quant_quantity,
+    SUM(quantity_in) AS quantity_in, SUM(quantity_out) AS quantity_out,
+    SUM(quantity_error) AS quantity_error
+    FROM quant_summary qs
+    JOIN stock_location sl
+    ON qs.location_id = sl.id
+    WHERE quantity_error > 0.01
+    AND sl.usage = 'internal'
+    GROUP BY qs.location_id, qs.product_id, qs.package_id, qs.owner_id, qs.lot_id
+    ORDER BY qs.location_id, product_id, qs.package_id, qs.owner_id, qs.lot_id
+"""
+
+RESERVED_STATEMENT = """
+SELECT COALESCE(SUM(product_qty), 0.0) AS reserved_in_moves
+ FROM stock_move_line
+ WHERE location_id = %(location_id)s
+   AND product_id = %(product_id)s
+   AND COALESCE(lot_id, 0) = %(lot_id)s
+   AND COALESCE(package_id, 0) = %(package_id)s
+   AND COALESCE(owner_id, 0) = %(owner_id)s
+"""
+
+
+class StockQuant(models.Model):
+    _inherit = "stock.quant"
+
+    quantity_error = fields.Float(
+        "Quantity in Error",
+        help="Difference between computed and actual quantity",
+        readonly=True,
+    )
+    quantity_computed = fields.Float(
+        "Quantity in Stock Move Lines",
+        help="Difference between computed and actual quantity",
+        readonly=True,
+    )
+
+    @api.constrains("product_id", "reserved_quantity")
+    def check_reserved_quantity(self):
```

Comment:
The problem here is that this constraint method is called before `stock.move.line` is updated, therefore it is risen any time we perform` action_assign()` in  a stock picking. More specifically, the aforementioned action calls

https://github.com/OCA/OCB/blob/13.0/addons/stock/models/stock_move.py#L1123

and a call to `self.env['stock.quant']._update_reserved_quantity()` therein.

Reply:
Good catch. Please feel free to change anything that is needed. You can fork my branch and then put a PR on my repo....

Reply:
I am trying to think of a use case where having this constraint will be useful. I am not sure, can you please elaborate on that decision? In my opinion, it is not needed. What we might need is an action that finds faulty quants/stock move lines, and a cron job that corrects them, with proper logging etc. The action that finds faulty stuff is already there, it is the constraint method. I can implement a correction method, that would basically equate `quant.reserved_quantity` and `sum(move_lines.product_qty`

Reply:
The thing is that with a repair action, without a constraint, you can never be sure where the inconsistencies come from. When an exception is thrown the moment the database threatens to be made inconsistent, you know from the stack trace exactly where the error is coming from.